helix-core/src/main/java/org/apache/helix/spectator/RoutingTable.java

```diff
@@ -30,53 +30,80 @@
+ * {resource, partition, state} to {instances} map function.
  */
 class RoutingTable {
   private static final Logger logger = LoggerFactory.getLogger(RoutingTable.class);
 
   // mapping a resourceName to the ResourceInfo
   private final Map<String, ResourceInfo> _resourceInfoMap;
+
   // mapping a resource group name to a resourceGroupInfo
   private final Map<String, ResourceGroupInfo> _resourceGroupInfoMap;
 
   private final Collection<LiveInstance> _liveInstances;
   private final Collection<InstanceConfig> _instanceConfigs;
   private final Collection<ExternalView> _externalViews;
+  private final Collection<CustomizedView> _customizedViews;
+
+  private final PropertyType _propertyType;
+  private final String _aggregationType;
 
   public RoutingTable() {
     this(Collections.<ExternalView> emptyList(), Collections.<InstanceConfig> emptyList(),
         Collections.<LiveInstance> emptyList());
   }
 
+  public RoutingTable(PropertyType propertyType, String aggregationType) {
+    this(Collections.<ExternalView> emptyList(), Collections.<CustomizedView> emptyList(), Collections.<InstanceConfig> emptyList(),
+        Collections.<LiveInstance> emptyList(), propertyType, aggregationType);
+  }
+
   public RoutingTable(Map<String, Map<String, Map<String, CurrentState>>> currentStateMap,
       Collection<InstanceConfig> instanceConfigs, Collection<LiveInstance> liveInstances) {
     // TODO Aggregate currentState to an ExternalView in the RoutingTable, so there is no need to
     // refresh according to the currentStateMap. - jjwang
-    this(Collections.<ExternalView> emptyList(), instanceConfigs, liveInstances);
+    this(Collections.<ExternalView> emptyList(), Collections.<CustomizedView> emptyList(),
+        instanceConfigs, liveInstances, PropertyType.CURRENTSTATES, null);
     refresh(currentStateMap);
   }
 
   public RoutingTable(Collection<ExternalView> externalViews,
       Collection<InstanceConfig> instanceConfigs, Collection<LiveInstance> liveInstances) {
+    this(externalViews, Collections.<CustomizedView> emptyList(), instanceConfigs, liveInstances,
+        PropertyType.EXTERNALVIEW, null);
+  }
+
+  protected RoutingTable(Collection<ExternalView> externalViews,
+      Collection<CustomizedView> customizedViews, Collection<InstanceConfig> instanceConfigs,
+      Collection<LiveInstance> liveInstances, PropertyType propertytype, String aggregationType) {
+    _propertyType = propertytype;
+    _aggregationType = aggregationType;
     _resourceInfoMap = new HashMap<>();
     _resourceGroupInfoMap = new HashMap<>();
     _liveInstances = new HashSet<>(liveInstances);
     _instanceConfigs = new HashSet<>(instanceConfigs);
     _externalViews = new HashSet<>(externalViews);
-    refresh(externalViews);
+    _customizedViews = new HashSet<>(customizedViews);
+    if (_propertyType == PropertyType.CUSTOMIZEDVIEW) {
+      refreshCustomizedView(customizedViews);
+    } else {
```

Comment:
can we specify propertyType in case we'd like to extend later or the wrong type is input.

Reply:
Add a general function that has switch case. I think that satisfies what you are proposing.